File: src/Serverlife/Core/OriginOverrideStore.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Serverlife.Core;

/// <summary>
/// User corrections to OriginClassifier's folder-based guess, persisted so a pin survives
/// a restart. Keyed by working directory when one is known — the same signal the
/// classifier itself uses — and by process name when it isn't, since that's all there is
/// to go on for something like a service whose folder can't be read (see the GT3 case:
/// WorkingDirectory unreadable, so it fell back to showing under Mine as Unknown).
/// </summary>
public static class OriginOverrideStore
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true,
        Converters = { new JsonStringEnumConverter() },
    };

    public static string StorePath { get; set; } = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "Serverlife", "origin-overrides.json");

    private static readonly Dictionary<string, ServerOrigin> Overrides = Load();

    public static string KeyFor(string? workingDirectory, string processName) =>
        !string.IsNullOrWhiteSpace(workingDirectory)
            ? "dir:" + workingDirectory.TrimEnd('\\').ToLowerInvariant()
            : "proc:" + Path.GetFileNameWithoutExtension(processName).ToLowerInvariant();

    public static ServerOrigin? Get(string key) =>
        Overrides.TryGetValue(key, out var origin) ? origin : null;

    public static void Set(string key, ServerOrigin origin)
    {
        Overrides[key] = origin;
        Save();
    }

    public static void Clear(string key)
    {
        if (Overrides.Remove(key))
            Save();
    }

    private static Dictionary<string, ServerOrigin> Load()
    {
        try
        {
            if (File.Exists(StorePath))
            {
                var json = File.ReadAllText(StorePath);
                if (JsonSerializer.Deserialize<Dictionary<string, ServerOrigin>>(json, JsonOptions) is { } loaded)
                    return loaded;
            }
        }
        catch
        {
            // Corrupt or unreadable overrides fall back to none rather than blocking startup.
        }
        return new Dictionary<string, ServerOrigin>();
    }

    private static void Save()
    {
        try
        {
            var dir = Path.GetDirectoryName(StorePath)!;
            Directory.CreateDirectory(dir);
            File.WriteAllText(StorePath, JsonSerializer.Serialize(Overrides, JsonOptions));
        }
        catch
        {
            // Best-effort; never block the app on this.
        }
    }
}

File: macos/Sources/ServerlifeCore/OriginOverrideStore.swift
import Foundation

/// User corrections to OriginClassifier's folder-based guess, persisted so a pin survives
/// a restart. Keyed by working directory when one is known — the same signal the
/// classifier itself uses — and by process name when it isn't, since that's all there is
/// to go on for something whose folder can't be read (a service running under another
/// uid, say).
public enum OriginOverrideStore {
    public static var storeURLOverride: URL?

    private static var storeURL: URL {
        if let storeURLOverride { return storeURLOverride }
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        return appSupport.appendingPathComponent("Serverlife", isDirectory: true)
            .appendingPathComponent("origin-overrides.json")
    }

    private static var cache: [String: ServerOrigin]?

    public static func key(workingDirectory: String?, processName: String) -> String {
        if let workingDirectory, !workingDirectory.trimmingCharacters(in: .whitespaces).isEmpty {
            let trimmed = workingDirectory.hasSuffix("/") ? String(workingDirectory.dropLast()) : workingDirectory
            return "dir:" + trimmed.lowercased()
        }
        let bare = (processName as NSString).deletingPathExtension
        return "proc:" + bare.lowercased()
    }

    public static func get(_ key: String) -> ServerOrigin? {
        loaded()[key]
    }

    public static func set(_ key: String, origin: ServerOrigin) {
        var overrides = loaded()
        overrides[key] = origin
        cache = overrides
        save(overrides)
    }

    public static func clear(_ key: String) {
        var overrides = loaded()
        guard overrides.removeValue(forKey: key) != nil else { return }
        cache = overrides
        save(overrides)
    }

    /// Re-reads from disk on first use; a test can reset `storeURLOverride` and then
    /// this without restarting the process.
    public static func resetCache() { cache = nil }

    private static func loaded() -> [String: ServerOrigin] {
        if let cache { return cache }
        let loaded = load()
        cache = loaded
        return loaded
    }

    private static func load() -> [String: ServerOrigin] {
        guard let data = try? Data(contentsOf: storeURL),
              let decoded = try? JSONDecoder().decode([String: ServerOrigin].self, from: data) else {
            // Corrupt or unreadable overrides fall back to none rather than blocking startup.
            return [:]
        }
        return decoded
    }

    private static func save(_ overrides: [String: ServerOrigin]) {
        do {
            let dir = storeURL.deletingLastPathComponent()
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            let data = try encoder.encode(overrides)
            try data.write(to: storeURL, options: .atomic)
        } catch {
            // Best-effort; never block the app on this.
        }
    }
}
